File: 0373-find-k-pairs-with-smallest-sums/0373-find-k-pairs-with-smallest-sums.py
import heapq

class Solution:
    def kSmallestPairs(self, nums1: List[int], nums2: List[int], k: int) -> List[List[int]]:
        n = len(nums1)
        m = len(nums2)
        
        h = []
        answer = []
        
        heapq.heappush(h, (nums1[0] + nums2[0], (0, 0)))
        
        visited = set()
        
        while len(answer) < k and h:
            s, (i, j) = heapq.heappop(h)
            answer.append((nums1[i], nums2[j]))
            
            if i + 1 < n and (i + 1, j) not in visited:
                visited.add((i + 1, j))
                heapq.heappush(h, (nums1[i + 1] + nums2[j], (i + 1, j)))
                
            if j + 1 < m and (i, j + 1) not in visited:
                visited.add((i, j + 1))
                heapq.heappush(h, (nums1[i] + nums2[j + 1], (i, j + 1)))

        return answer   

File: 0373-find-k-pairs-with-smallest-sums/0373-find-k-pairs-with-smallest-sums.rs
use std::collections::{BinaryHeap, HashSet};

impl Solution {
    pub fn k_smallest_pairs(nums1: Vec<i32>, nums2: Vec<i32>, k: i32) -> Vec<Vec<i32>> {
        let n = nums1.len();
        let m = nums2.len();
        let mut k = k;
        
        let mut h: BinaryHeap<(i32, (i32, i32))> = BinaryHeap::new();
        h.push((-(nums1[0] + nums2[0]), (0, 0)));
        
        let mut visited: HashSet<(i32, i32)> = HashSet::new();

        let mut answer: Vec<Vec<i32>> = vec![];
        
        while k > 0 && !h.is_empty() {
            let (sum, (i, j)) = h.pop().unwrap();
            answer.push(vec![nums1[i as usize], nums2[j as usize]]);
            
            if ((i + 1) as usize) < n && !visited.contains(&(i + 1, j)) {
                visited.insert((i + 1, j));
                h.push((-(nums1[(i + 1) as usize] + nums2[j as usize]), (i + 1, j)));
            }
            
            if ((j + 1) as usize) < m && !visited.contains(&(i, j + 1)) {
                visited.insert((i, j + 1));
                h.push((-(nums1[i as usize] + nums2[(j + 1) as usize]), (i , j + 1)));
            }
            k -= 1;
        }
        return answer;
    }
}